Added service configuration demo

kawsark opened the pull request:
Adding service configuration demo from repo: [https://github.com/kawsark/thomas_cc_demo]. Made the following updates per Roger's review:
- Added version information for Packer, Terraform, Consul and Vault.
- Updated product service to use logger.debug()
- Added more explanation of how Envconsul retrieves credentials from Vault and key/value configuration from Consul.
- Removed need for listing_wrapper.sh script, Envconsul will call /usr/bin/node directly to start the Node.js application.
- Setup VAULT_TOKEN and VAULT_ADDR during Vault server initialization.

rberlind commented:
I suspect there are some files in packer directory, especially those ending with "~" and others starting with "." that could have been excluded in .gitignore.  Apart from temporary or other files that would not be needed in any case, are there also files here left over from Thomas's original demo not needed by your Packer image?  If so, those could be deleted too.

rberlind commented:
In step 9 of provisioning, you might also want to mention `curl -s http://whatismyip.akamai.com` which is in a comment in terraform.auto.tfvars.example.

There is really no need to run `terraform plan` since running `terraform apply` does that.

rberlind commented:
For service configuration section, you might want to mention where to find the configuration (that is, under the Key/Value menu and then under the <service>/config path.

kawsark commented:
Thanks @rberlind for the detailed feedback. The last commit includes the following set of changes:
1. Removed references to other use-cases. Made architecture overview more concise and updated diagram with Vault. My preference is to keep the 2 READMEs separate. 
2. Removed all references to connect vs. no-connect from README, terraform code and packer
3. Updated packer directory to remove *~ and .built files. Removed GCP builder in packer.
4. Removed Makefile and replaced with a simple packer.example.sh. Tested by creating a set of new AMIs in us-west-1 region.
5. Updated Terraform code to ensure there are no resource name conflicts. Tested creating duplicate stacks using terraform OSS workspaces in the same region, with only difference being `project_name` variable.
6. Added information regarding systemd unit file for both services `product.service` and `listing.service`.
7. Updated git clone URL to consul-guides. **Note: there will be a follow-up PR to remove the PR only step: `cd consul-guides && git fetch && git checkout add-service-configuration && cd ..`**

kawsark commented:
8. Also updated README.md with above feedback.

rberlind commented:
I've actually made most of the  changes suggested below for the two README.md files in a separate branch (add-service-configuration-roger) so you can see what they would look like and copy them into your PR if you want.  But I did not want to force them upon you by including them on this branch.

I'm nit-picking here, @kawsark, but you should be using "envconsul" and "Consul Template", at least in the middle of sentences, based on what the REAME.md files for those solutions do.  Using "Envconsul" at the beginning of a sentence is ok.

I think the reference to the AWS Secrets engine should be to the MongoDB Database secrets engine.

It would be good to fill in more details in https://github.com/hashicorp/consul-guides/blob/add-service-configuration/service-configuration/service-configuration-demo/application/listing-service/README.md

Under section "Terraform steps" section in the "Running the Demo" doc, there is an issue with the formatting of the paragraph that starts with "Replace `<your access key ID>` with your AWS Access Key ID ".  For some reason, it is being treated as if it is inside back-ticks. Perhaps because you indented the paragraph?  The result is that the hyperlink you tried to create for "[Access Keys (Access Key ID and Secret Access Key)]" was not created.

At the beginning of the "Service Configuration for Listing Service" section, "exit out of the Listing SSH session" should be "exit out of the Product SSH session".

I suggest the following header changes:
- "Service Configuration for Product service" to "Service Configuration for Product Service with Consul Template"
- "Service Configuration for Listing service" to "Service Configuration for Listing Service with Envconsul"
- "Dynamic Credentials for Listing service" to "Dynamic Credentials for Listing Service with Envconsul"
- "Dynamic Credentials for Product service" to "Dynamic Credentials for Product Service with Vault API"
- And then add these various sub-headers to the "Demo Steps" section at the top.
- Finally, the "Secrets Management" item under "Demo Steps" should be "Dynamic Credentials" to match what you have further down.

rberlind commented:
Also, there are still git clones being done against Thomas's repos.  I think you plan on removing that after the first PR is approved.  Right?
```
$grep -R "thomashashi" *
packer/files/install_node.sh:git clone https://github.com/thomashashi/listing-service.git
packer/files/install_products.sh:git clone https://github.com/thomashashi/product-service.git
packer/files/install_webclient.sh:git clone https://github.com/thomashashi/simple-client.git
```

kawsark commented:
Thank you @rberlind, these are all fixed on the latest commit.

> Also, there are still git clones being done against Thomas's repos. I think you plan on removing that after the first PR is approved. Right?
> 
> ```
> $grep -R "thomashashi" *
> packer/files/install_node.sh:git clone https://github.com/thomashashi/listing-service.git
> packer/files/install_products.sh:git clone https://github.com/thomashashi/product-service.git
> packer/files/install_webclient.sh:git clone https://github.com/thomashashi/simple-client.git
> ```
